h5py/_conv.pyx

```diff
@@ -814,15 +815,19 @@ cdef int conv_ndarray2vlen(void* ipt,
         cnp.ndarray ndarray
         size_t len, nbytes
         PyObject* buf_obj0
+        Py_buffer view
 
     buf_obj0 = buf_obj[0]
-    ndarray = np.ascontiguousarray(<cnp.ndarray> buf_obj0)
+    ndarray = <cnp.ndarray> buf_obj0
     len = ndarray.shape[0]
     nbytes = len * max(outtype.get_size(), intype.get_size())
 
     data = emalloc(nbytes)
 
-    memcpy(data, ndarray.data, intype.get_size() * len)
+    PyObject_GetBuffer(ndarray, &view, PyBUF_INDIRECT)
```

Comment:
INDIRECT is the opposite extreme to SIMPLE ([C API buffer docs](https://docs.python.org/3/c-api/buffer.html#shape-strides-suboffsets)) - if I've understood this correctly, we need at least STRIDES here to accept a non-contiguous array, and then we copy it to contiguous memory in the next step.

I think/hope this function only gets used with 1D arrays; I suspect it would go wrong if it got a multidimensional array.

Reply:
Thanks!
This is the same code I had, but I only tried `SIMPLE`, `ND`. I had not considered going further (not sure why).